### PR TITLE
TSV-981: Validate post code

### PR DESCRIPTION
**Pull request by steve-edmans**

### CODE REVIEW

File: app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsFormat.scala [hunk 1]
```diff
@@ -46,7 +47,11 @@ object PersonalDetailsFormat {
         case Some(nino) => Some(Nino(nino))
         case _ => None
       } and
-        (__ \ "postCode").readNullable[String]
+        (__ \ "postCode").readNullable[String].filter(ValidationError("invalid postcode format")){
+          case postCode if !postCode.isDefined => true
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsFormat.scala]: I think this pattern match might read better 
```
    case None => true
    case Some(postCodeValidation(_)) => true
    case _ => false
```

File: test/generators/PostCodeGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    "TF", "TN", "TQ", "TR", "TS", "TW", "UB", "W", "WA", "WC", "WD", "WF", "WN", "WR", "WS", "WV", "YO", "ZE")
+
+  private val outboundOptions: Map[String, List[String]] = Map(
+    "WC" -> List("1A", "1B", "1E", "1H", "1N", "1R", "1V", "1X", "2A", "2B", "2E", "2H", "2N", "2R"),
+    "EC" -> List("1A", "1M", "1N", "1P", "1R", "1V", "1Y", "2A", "2M", "2N", "2P", "2R", "2V", "2Y", "3A", "3M", "3N",
+      "3P", "3R", "3V", "4A", "4M", "4N", "4P", "4R", "4V", "4Y")
+  )
+
+  private val invalidCharacters : List[Char] = List('C', 'I', 'K', 'M', 'O', 'V', 'c', 'i', 'k', 'm', 'o', 'v')
+  private val postCodeArea : Gen[String] = oneOf(postCodeAreas)
```
Comment (anshulbajpai) [inline test/generators/PostCodeGenerator.scala]: Probably can inline at point of usage

File: test/generators/PostCodeGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  private val unitDigitsAllowedUpper = "ABDEFGHJLNPQRSTUWXYZ".toCharArray
+  private val unitDigitsAllowedLower = unitDigitsAllowedUpper.map(_.toLower)
+  private val unitDigit = Gen.frequency((1, oneOf(unitDigitsAllowedLower)), (5, oneOf(unitDigitsAllowedUpper)))
+
+  private def outbound(area: String): Gen[String] = {
+    area match {
```
Comment (anshulbajpai) [inline test/generators/PostCodeGenerator.scala]: Probably can write like this 
`outboundOptions.get(area).map(oneOf).getOrElse(Gen.choose(1, 99).map(_.toString))`

File: test/generators/PostCodeGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      case _ if outboundOptions.keySet.contains(area) => oneOf(outboundOptions(area))
+      case _ => Gen.choose(1, 99).map(_.toString)
+    }
+  }
+
+  private def capitalizeAreaCharacter(area: String, position: Int, randomNumber: Int) : String = {
+    area.length match {
```
Comment (anshulbajpai) [inline test/generators/PostCodeGenerator.scala]: Can we not create the random number inside this method?

Comment (anshulbajpai) [inline test/generators/PostCodeGenerator.scala]: I see we are using pattern matching but there is actually no pattern matching happening here.
Why can't we use the simple if-else expression here?

File: test/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsValidationFormatSpec.scala [hunk 5]
```diff
@@ -36,6 +38,81 @@ class PersonalDetailsValidationFormatSpec
...
+          personalDetails.lastName,
+          personalDetails.dateOfBirth,
+          " " + postCode
+        ).toJson.toString()
+        an [JsResultException] should be thrownBy
+          Json.parse(personalDetailsWithPostCodeJson).as[PersonalDetails]
+      }
+    }
+
+    "correctly fail to parse valid postcodes that do not end the field" in new Setup {
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsValidationFormatSpec.scala]: We can trim the input string

File: test/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsValidationFormatSpec.scala [hunk 6]
```diff
@@ -36,6 +38,81 @@ class PersonalDetailsValidationFormatSpec
...
+          postCode + " " + secondPostCode
+        ).toJson.toString()
+        an [JsResultException] should be thrownBy
+          Json.parse(personalDetailsWithPostCodeJson).as[PersonalDetails]
+      }
+    }
+
+    "fail validation for invalid postcodes" in new Setup {
+      import PersonalDetailsFormat._
+      val invalidPostCode = listOfN[Char](7, frequency((1,choose(65.toChar, 90.toChar)), (1, choose(97.toChar, 122.toChar)))).map(_.mkString)
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsValidationFormatSpec.scala]: Can you see if it is possible to create an alphanumeric string which does not satisfy postcode generator?

File: app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsFormat.scala [hunk 7]
```diff
@@ -30,6 +30,7 @@ import scala.util.Try
 object PersonalDetailsFormat {
   implicit val withNinoFormats: Format[PersonalDetailsWithNino] = Json.format[PersonalDetailsWithNino]
   implicit val withPostCodeFormats: Format[PersonalDetailsWithPostCode] = Json.format[PersonalDetailsWithPostCode]
+  private val postCodeValidation = """^([A-Za-z][A-HJ-Ya-hj-y]?[0-9][A-Za-z0-9]?|[A-Za-z][A-HJ-Ya-hj-y][A-Za-z])\s?[0-9][ABDEFGHJLNPQRSTUWXYZabdefghjlnpqrstuwxyz]{2}$""".r
```
Comment (burakbala) [inline app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsFormat.scala:33]: If we do toUppercase before passing the value to the regex, we can have a slightly less complicated looking regex:

"""^([A-Z][A-HJ-Y]?[0-9][A-Z0-9]?|[A-Z][A-HJ-Y][A-Z])\s?[0-9][ABDEFGHJLNPQRSTUWXYZ]{2}$""".r

Comment (steve-edmans) [reply app/uk/gov/hmrc/personaldetailsvalidation/formats/PersonalDetailsFormat.scala:33]: Do we really want to perform two actions when one will work, even if the regular expression is a little bit more complicated?

**Comment by anshulbajpai** [comment]
👍 
@burakbala - I guess you had a comment. If you are happy then just merge it

**Pull request merged**